engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/SelectAndViewAnalyzer.java

```diff
@@ -350,100 +376,208 @@ private static boolean shouldPreserve(final SelectColumn sc) {
-            UpdateGraph updateGraph, RowSet parentRowset, String name, SelectColumn sc, WritableColumnSource<?> cs,
-            WritableColumnSource<?> underlyingSource, String[] parentColumnDependencies, ModifiedColumnSet mcsBuilder,
-            boolean isRedirected, boolean flattenResult, boolean alreadyFlattened) {
-        return new SelectColumnLayer(updateGraph, parentRowset, this, name, sc, cs, underlyingSource,
-                parentColumnDependencies,
-                mcsBuilder, isRedirected, flattenResult, alreadyFlattened);
     }
 
-    private SelectAndViewAnalyzer createLayerForConstantView(String name, SelectColumn sc, WritableColumnSource<?> cs,
-            String[] parentColumnDependencies, ModifiedColumnSet mcsBuilder, boolean flattenResult,
-            boolean alreadyFlattened) {
-        return new ConstantColumnLayer(this, name, sc, cs, parentColumnDependencies, mcsBuilder, flattenResult,
-                alreadyFlattened);
-    }
+    void addLayer(final Layer layer) {
+        layers.add(layer);
 
-    private SelectAndViewAnalyzer createLayerForView(String name, SelectColumn sc, ColumnSource<?> cs,
-            String[] parentColumnDependencies, ModifiedColumnSet mcsBuilder) {
-        return new ViewColumnLayer(this, name, sc, cs, parentColumnDependencies, mcsBuilder);
+        for (final String columnName : layer.getLayerColumnNames()) {
+            columnToLayerIndex.put(columnName, layer.getLayerIndex());
+        }
     }
 
-    private SelectAndViewAnalyzer createLayerForPreserve(String name, SelectColumn sc, ColumnSource<?> cs,
-            String[] parentColumnDependencies, ModifiedColumnSet mcsBuilder) {
-        return new PreserveColumnLayer(this, name, sc, cs, parentColumnDependencies, mcsBuilder);
+    public int getNextLayerIndex() {
+        return layers.size();
     }
 
-    abstract void populateModifiedColumnSetRecurse(ModifiedColumnSet mcsBuilder, Set<String> remainingDepsToSatisfy);
+    public static abstract class Layer implements LogOutputAppendable {
+
+        static final int BASE_LAYER_INDEX = 0;
+        static final int REDIRECTION_LAYER_INDEX = 1;
+
+        enum GetMode {
+            All, New, Published
+        }
+
+        /**
+         * The layerIndex is used to identify each layer uniquely within the bitsets for completion.
+         */
+        private final int layerIndex;
+
+        public Layer(int layerIndex) {
+            this.layerIndex = layerIndex;
+        }
+
+        int getLayerIndex() {
+            return layerIndex;
+        }
+
+        @Override
+        public String toString() {
+            return new LogOutputStringImpl().append(this).toString();
+        }
 
-    enum GetMode {
-        All, New, Published
+        public void startTrackingPrev() {
+            // default is that there is nothing to do
+        }
+
+        abstract Set<String> getLayerColumnNames();
+
+        abstract void populateModifiedColumnSetInReverse(
```

Comment:
I think this can be done iteratively from base to final layer more efficiently/clearly. We think maybe as layers are added, we can map name to source deps.